Various fixes around dispatching

guusdk opened the pull request:
This PR has various fixes that were needed to add more than just the singlular RoundRobin dispatcher. The end result is the addition of a new Dispatcher, one that sends an offer to all available agents.

guusdk commented:
Although the admin console modifications work, I've not tested this with actual agents yet. That should be done before merging/releasing this.

deleolajide commented:
Exception in thread "Dispatch offer - queue: Default Queue" java.lang.NullPointerException
        at org.jivesoftware.xmpp.workgroup.dispatcher.AbstractDispatcher.lambda$findAgentSessionByName$0(AbstractDispatcher.java:491)
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)
        at java.util.concurrent.ConcurrentHashMap$KeySpliterator.tryAdvance(ConcurrentHashMap.java:3535)
        at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
        at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:464)
        at org.jivesoftware.xmpp.workgroup.dispatcher.AbstractDispatcher.findAgentSessionByName(AbstractDispatcher.java:492)
        at org.jivesoftware.xmpp.workgroup.dispatcher.AbstractDispatcher.dispatch(AbstractDispatcher.java:176)
        at org.jivesoftware.xmpp.workgroup.dispatcher.AbstractDispatcher$2.run(AbstractDispatcher.java:140)
Security framework of XStream not explicitly initialized, using predefined black list on your own risk.
Exception in thread "Dispatch offer - queue: Default Queue" java.lang.NullPointerException
        at org.jivesoftware.xmpp.workgroup.dispatcher.AbstractDispatcher.lambda$findAgentSessionByName$0(AbstractDispatcher.java:491)
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)
        at java.util.concurrent.ConcurrentHashMap$KeySpliterator.tryAdvance(ConcurrentHashMap.java:3535)
        at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
        at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:464)
        at org.jivesoftware.xmpp.workgroup.dispatcher.AbstractDispatcher.findAgentSessionByName(AbstractDispatcher.java:492)
        at org.jivesoftware.xmpp.workgroup.dispatcher.AbstractDispatcher.dispatch(AbstractDispatcher.java:176)
        at org.jivesoftware.xmpp.workgroup.dispatcher.AbstractDispatcher$2.run(AbstractDispatcher.java:140)
Exception in thread "Dispatch offer - queue: Default Queue" java.lang.NullPointerException
        at org.jivesoftware.xmpp.workgroup.dispatcher.AbstractDispatcher.lambda$findAgentSessionByName$0(AbstractDispatcher.java:491)
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)
        at java.util.concurrent.ConcurrentHashMap$KeySpliterator.tryAdvance(ConcurrentHashMap.java:3535)
        at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
        at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:464)
        at org.jivesoftware.xmpp.workgroup.dispatcher.AbstractDispatcher.findAgentSessionByName(AbstractDispatcher.java:492)
        at org.jivesoftware.xmpp.workgroup.dispatcher.AbstractDispatcher.dispatch(AbstractDispatcher.java:176)
        at org.jivesoftware.xmpp.workgroup.dispatcher.AbstractDispatcher$2.run(AbstractDispatcher.java:140)
Exception in thread "Dispatch offer - queue: Default Queue" java.lang.NullPointerException
        at org.jivesoftware.xmpp.workgroup.dispatcher.AbstractDispatcher.lambda$findAgentSessionByName$0(AbstractDispatcher.java:491)
        at java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:174)
        at java.util.concurrent.ConcurrentHashMap$KeySpliterator.tryAdvance(ConcurrentHashMap.java:3535)
        at java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:126)
        at java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:498)
        at java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:485)
        at java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:471)
        at java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:152)
        at java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
        at java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:464)
        at org.jivesoftware.xmpp.workgroup.dispatcher.AbstractDispatcher.findAgentSessionByName(AbstractDispatcher.java:492)
        at org.jivesoftware.xmpp.workgroup.dispatcher.AbstractDispatcher.dispatch(AbstractDispatcher.java:176)
        at org.jivesoftware.xmpp.workgroup.dispatcher.AbstractDispatcher$2.run(AbstractDispatcher.java:140)

guusdk commented:
Thanks @deleolajide - I could easily reproduce that. I've now fixed that NullPointerException.

I did some more testing, which made some more problems bubble up. I have fixed those, and did some basic testing with the 'Round Robin' dispatcher, which now didn't throw exceptions.

The 'Opportunistic' dispatcher also worked as intended up to the point where I was able to test. For agents, I was using two Spark clients. Those both received an offer, but displaying those offers was problematic. That's probably because I was running Spark twice on one laptop. Would you mind giving it another go?

deleolajide commented:
I can confirm that the NPE exception is fixed. However, I am testing with two agents using the vanilla converse client with the fastpath chatbot in a chat conversation to join and accept offers. There is an exception happening in this context
```
2021.01.13 10:05:05 org.jivesoftware.xmpp.workgroup.chatbot.Chatbot - acceptRejectOffer failed
org.jivesoftware.util.NotFoundException: null
	at org.jivesoftware.xmpp.workgroup.request.Request.getRequest(Request.java:96) ~[fastpath-4.5.0-SNAPSHOT.jar!/:?]
	at org.jivesoftware.xmpp.workgroup.chatbot.Chatbot.acceptRejectOffer(Chatbot.java:454) [fastpath-4.5.0-SNAPSHOT.jar!/:?]
	at org.jivesoftware.xmpp.workgroup.chatbot.Chatbot.onMessage(Chatbot.java:216) [fastpath-4.5.0-SNAPSHOT.jar!/:?]
	at org.jivesoftware.xmpp.workgroup.MessageHandler.process(MessageHandler.java:55) [fastpath-4.5.0-SNAPSHOT.jar!/:?]
	at org.jivesoftware.xmpp.workgroup.Workgroup.process(Workgroup.java:460) [fastpath-4.5.0-SNAPSHOT.jar!/:?]
	at org.jivesoftware.xmpp.workgroup.Workgroup.process(Workgroup.java:473) [fastpath-4.5.0-SNAPSHOT.jar!/:?]
	at org.jivesoftware.xmpp.workgroup.WorkgroupManager.processPacket(WorkgroupManager.java:683) [fastpath-4.5.0-SNAPSHOT.jar!/:?]
	at org.jivesoftware.openfire.component.InternalComponentManager$RoutableComponents.process(InternalComponentManager.java:859) [xmppserver-4.6.0.jar:4.6.0]
	at org.jivesoftware.openfire.spi.RoutingTableImpl.routeToComponent(RoutingTableImpl.java:451) [xmppserver-4.6.0.jar:4.6.0]
	at org.jivesoftware.openfire.spi.RoutingTableImpl.routePacket(RoutingTableImpl.java:266) [xmppserver-4.6.0.jar:4.6.0]
	at org.jivesoftware.openfire.MessageRouter.route(MessageRouter.java:134) [xmppserver-4.6.0.jar:4.6.0]
	at org.jivesoftware.openfire.spi.PacketRouterImpl.route(PacketRouterImpl.java:79) [xmppserver-4.6.0.jar:4.6.0]
	at org.jivesoftware.openfire.SessionPacketRouter.route(SessionPacketRouter.java:111) [xmppserver-4.6.0.jar:4.6.0]
	at org.jivesoftware.openfire.SessionPacketRouter.route(SessionPacketRouter.java:66) [xmppserver-4.6.0.jar:4.6.0]
	at org.jivesoftware.openfire.websocket.StreamManagementPacketRouter.route(StreamManagementPacketRouter.java:61) [xmppserver-4.6.0.jar:4.6.0]
	at org.jivesoftware.openfire.websocket.XmppWebSocket.processStanza(XmppWebSocket.java:242) [xmppserver-4.6.0.jar:4.6.0]
	at org.jivesoftware.openfire.websocket.XmppWebSocket.onTextMethod(XmppWebSocket.java:116) [xmppserver-4.6.0.jar:4.6.0]
	at sun.reflect.GeneratedMethodAccessor51.invoke(Unknown Source) ~[?:?]
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[?:1.8.0_144]
	at java.lang.reflect.Method.invoke(Method.java:498) ~[?:1.8.0_144]
	at org.eclipse.jetty.websocket.common.events.annotated.CallableMethod.call(CallableMethod.java:70) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.websocket.common.events.annotated.OptionalSessionCallableMethod.call(OptionalSessionCallableMethod.java:72) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.websocket.common.events.JettyAnnotatedEventDriver.onTextMessage(JettyAnnotatedEventDriver.java:248) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.websocket.common.message.SimpleTextMessage.messageComplete(SimpleTextMessage.java:69) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.websocket.common.events.AbstractEventDriver.appendMessage(AbstractEventDriver.java:65) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.websocket.common.events.JettyAnnotatedEventDriver.onTextFrame(JettyAnnotatedEventDriver.java:240) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.websocket.common.events.AbstractEventDriver.incomingFrame(AbstractEventDriver.java:150) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.websocket.common.WebSocketSession.incomingFrame(WebSocketSession.java:326) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.websocket.common.extensions.AbstractExtension.nextIncomingFrame(AbstractExtension.java:147) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.websocket.common.extensions.compress.PerMessageDeflateExtension.nextIncomingFrame(PerMessageDeflateExtension.java:112) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.websocket.common.extensions.compress.CompressExtension.forwardIncoming(CompressExtension.java:168) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.websocket.common.extensions.compress.PerMessageDeflateExtension.incomingFrame(PerMessageDeflateExtension.java:92) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.websocket.common.extensions.ExtensionStack.incomingFrame(ExtensionStack.java:202) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.websocket.common.Parser.notifyFrame(Parser.java:225) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.websocket.common.Parser.parseSingleFrame(Parser.java:259) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onFillable(AbstractWebSocketConnection.java:459) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.websocket.common.io.AbstractWebSocketConnection.onFillable(AbstractWebSocketConnection.java:440) [websocket-common-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311) [jetty-io-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105) [jetty-io-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.io.ssl.SslConnection$DecryptedEndPoint.onFillable(SslConnection.java:540) [jetty-io-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.io.ssl.SslConnection.onFillable(SslConnection.java:395) [jetty-io-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.io.ssl.SslConnection$2.succeeded(SslConnection.java:161) [jetty-io-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105) [jetty-io-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104) [jetty-io-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:336) [jetty-util-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:313) [jetty-util-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:171) [jetty-util-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:129) [jetty-util-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:375) [jetty-util-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:806) [jetty-util-9.4.31.v20200723.jar:9.4.31.v20200723]
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:938) [jetty-util-9.4.31.v20200723.jar:9.4.31.v20200723]
	at java.lang.Thread.run(Thread.java:748) [?:1.8.0_144]
```

guusdk commented:
Is this with the pre-existing Round Robin dispatcher, or with the new Opportunistic one (or both)? If it's the Opportunistic one, I can image that one agent accepting or rejecting an offer will have made the request unavailable, causing the second agent to reject it to  not be able to find it anymore.

deleolajide commented:
> or with the new Opportunistic one (or both)?

I forgot to mention which. I was testing with the Opportunistic dispatcher.

guusdk commented:
That was a problem in time-related calculations. I've now fixed that.

guusdk commented:
I've fixed most of the issues now, and tested with ChatBot (not with a non-ChatBot client yet).

One thing that stands out is #48 (but that issue is not introduced by this PR).

Another issue that I've observed is that things do not go well when a queue gets reconfigured to use a different dispatcher. That currently requires a restart of Openfire (perhaps just restarting the FastPath plugin also works). It's something that I'd like to test further.

guusdk commented:
@deleolajide as discussed, I've now added a new offline setting: 'redirect to muc' and removed this as a hardcoded default from the embedded converse.js client. The converse.js client will instead retrieve the offline settings when appropriate, and act accordingly. In that function, I've provided a basic implementation of the 'redirect to muc' strategy (to basically mimic the old, hard-coded behavior), but the other actions are left unimplemented. Here is where I need help: my ability to work with the Converse code simply isn't enough to finish the implementation. Would you mind having a look? I left "todo's" in the relevant spots in `ofmeet.js`.

Beware that my last commit adds to the database update script that I created in an earlier commit. If your environment has already executed that script, some manual database tinkering is needed.

deleolajide commented:
Saving the offline settings for a workgroup seems to have an issue. 
The values for **Redirect to Web Page** and **Redirect to Chat Room** are not being saved.

![image](https://user-images.githubusercontent.com/110731/104807076-cfbe4000-57d3-11eb-8ba0-5920c91716db.png)

guusdk commented:
That's how the existing code works: it only saves the values for the selected item. It's a bit awkward...

deleolajide commented:
> it only saves the values for the selected item. It's a bit awkward...

Don't you think we should fix it? It will be nice to know what current settings are

guusdk commented:
The currently active setting will be selected, and it's properties shown. All other properties (those of unselected options) might not be saved/shown. Your confusion alone is proof that this needs to be improved. In its current state, it should be usable though.

guusdk commented:
I have added some more fixes. Notably, a few for #54 that let the Converse client play nice with Spark (which needs a couple of undocumented metadata bits).

deleolajide reviewed (approved):
Thanks for all the changes. All acceptable to me.
